### ISSUE TITLE
implement app hash as merkle root of utxo

### ISSUE THREAD

**Issue by sbellem** [bigchaindb/bigchaindb#1865]
_This is an optimization for when time allows._
_Description and explanation will be added soon._

> The app should respond to the Commit request with a byte array, which is the deterministic state root of the application. It is included in the header of the next block. It can be used to provide easily verified Merkle-proofs of the state of the application.

> When the app or tendermint restarts, they need to sync to a common height. When an ABCI connection is first established, Tendermint will call Info on the Query connection. The response should contain the LastBlockHeight and LastBlockAppHash - the former is the last block for which the app ran Commit successfully, the latter is the response from that Commit.

**Comment by ttmc** [comment]
The hash of the app state depends on what you consider to be the "app state."

In Bitcoin, the UTXO set is maybe enough, but in BigchainDB, some people will use a CREATE transaction to describe the initial state of an "asset" (e.g. an empty journal with a title), and then they'll use TRANSFER transactions to transfer that asset to themselves, using the `metadata` of each TRANSFER transaction to update the state of the asset (e.g. a new entry in the journal). The current state would have to include _all_ the transactions in that sequence (i.e. the whole journal), not just the last one.

**Comment by ttmc** [comment]
Hmmm, thinking about this more, after chatting with @sbellem ... 

All the unspent transaction IDs, together, carry footprints of all the `asset.data` and `metadata` in transactions that led to them, because their `inputs` carry the transaction IDs of previous, spent transaction outputs (in the `fulfills` part of the `inputs`), and so on, all the way back to all the original CREATE transactions.

In other words, if you changed any `asset.data` or `metadata` anywhere in the transaction DAG, that would cause a change in the transaction ID of at least one unspent transaction output. Nice.

**Comment by ttmc** [comment]
There's a different issue, however: the information in the `fulfillment` values doesn't propagate forward in the transaction DAG to affect the state contained in the set of transactions with unspent outputs; see issue #1874
